### PR TITLE
Fix: Locally copied version of last published workfile is not incremented

### PR DESCRIPTION
**Pull request by kalisp**
### Fix 1
When copied, the local workfile version keeps the published version number, when it must be +1 to follow OP's naming convention.

### Fix 2
Local workfile version's name is built from anatomy. This avoids to get workfiles with their publish template naming.

### Fix 3
In the case a subset has at least two tasks with published workfiles, for example `Modeling` and `Rigging`, launching `Rigging` was getting the first one with the `next` and trying to find representations, therefore `workfileModeling` and trying to match the current `task_name` (`Rigging`) with the `representation["context"]["task"]["name"]` of a Modeling representation, which was ending up to a `workfile_representation` to `None`, and exiting the process.

Trying to find the `task_name` in the `subset['name']` fixes it.

### Fix 4
Fetch input dependencies of workfile.

Replacing https://github.com/ynput/OpenPype/pull/4102 for changes to bring this home.

### CODE REVIEW

File: openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py
Comment (BigRoy) [inline openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py:113]: Families should be a list right?
```suggestion
                or "workfile" in subset["data"].get("families", [])
```

File: openpype/modules/sync_server/sync_server.py
Comment (BigRoy) [inline openpype/modules/sync_server/sync_server.py:308]: How heavy is the `is_representation_on_site` check? I'd expect that it might make much more sense to prior to this just get the "expected files list" from the sync server and periodically just check whether the file exists on disk (avoiding any database calls, etc.)? And then maybe after do a database check if needed? I'm asking because 5 seconds seems quite long to be an interval checking at if this is used in e.g. a prelaunch hook where the user is actively waiting for a result to happen.

Also, this while loop might need a maximum timeout too?

File: openpype/modules/sync_server/sync_server.py
Comment (BigRoy) [inline openpype/modules/sync_server/sync_server.py:342]: When does this happen through the prelaunch hook? The prelaunch hook only continues download if there's no local workfiles yet, right? So why does it try incrementing against local workfiles? Or what am I missing?

File: openpype/modules/sync_server/sync_server.py
Comment (kalisp) [reply openpype/modules/sync_server/sync_server.py:308]: I don't think we can reliably test presence of a file on the disk as some OSs will create a file there immediately and file changes until is fully downloaded. We could query DB for file size and compare it against file metadata from the disk. But can we get them reliably from all OSs and would it be actually faster? I don't feel that this would happen too often and it is quick check against indexed `_id`.

Was thinking about maximum timeout too, but what is reasonable value? 2 minutes, 5 minutes, 10?

File: openpype/modules/sync_server/sync_server.py
Comment (kalisp) [reply openpype/modules/sync_server/sync_server.py]: Hm, I think this might not be necessary, removed it.
Maybe @Tilix4 will have an idea why it is there.

File: openpype/modules/sync_server/sync_server.py
Comment (kalisp) [reply openpype/modules/sync_server/sync_server.py:306]: Added check for failed downloads, it could happen that downloading fails, so there is a limit how many tries should be attempted.

File: openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py
Comment (BigRoy) [inline openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py]:
```suggestion
                "No subset for asset '{}' with id '{}'.".format(
```

File: openpype/modules/sync_server/sync_server.py
Comment (Tilix4) [reply openpype/modules/sync_server/sync_server.py]: I'm not sure the highlighted code is relevant. But the idea is that it takes the last version number and increments. This way, you don't get confused if you're downloading the `v003` for the first time, the opened workfile is a `v004`, and not a `v001`.

Comment (BigRoy) [reply openpype/modules/sync_server/sync_server.py]: > I'm not sure the highlighted code is relevant. But the idea is that it takes the last version number and increments. This way, you don't get confused if you're downloading the v003 for the first time, the opened workfile is a v004, and not a v001.

Yes, that's what I'd expect too - but the logic that was here originally compared it with versions on disk locally which doesn't trigger since the logic wouldn't start if there were local files.

File: openpype/modules/sync_server/sync_server_module.py
Comment (Tilix4) [inline openpype/modules/sync_server/sync_server_module.py:842]: Missing docstring of return

File: openpype/modules/sync_server/sync_server_module.py
Comment (Tilix4) [inline openpype/modules/sync_server/sync_server_module.py:1069]: Missing docstring of return

**Review by Tilix4 — CHANGES_REQUESTED** [review]

File: openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py
Comment (BigRoy) [inline openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py:129]: This could also be `max(workfile_representations, key=lambda d: d["context"]["version"])`

File: openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py
Comment (BigRoy) [inline openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py:146]: Should we at least log prior or after this that's it downloading a workfile representation?

**Review by BigRoy — DISMISSED** [review]
I've posted some small comments. Note that I'm not using sync server module myself so I haven't tested this and also can't comment on whether this works as expected for someone using the module.

I'm solely approving based on that I think the logic seems ok.
It feels a bit weird that the copy last published workfile launch hook is set for blender, photoshop, tvpaint and aftereffects but not for many other apps like Maya, Max, Nuke, Fusion and I wonder why that is.

I'm not sure if there already is, but I'd expect that to be customizable for the project or module maybe? And additionally have some documentation explaining the feature. I don't see documentation about this [here](https://openpype.io/docs/module_site_sync).

**Review by Tilix4 — CHANGES_REQUESTED** [review]
Tested and it doesn't work. It does not fetch the last published file if missing locally, it also stucks the loop and OP hangs.

Also I got this error constantly:
```
Traceback (most recent call last):
  File "/home/felix/Documents/Dev/OpenPype/openpype/modules/sync_server/sync_server.py", line 512, in sync_loop
    except asyncio.exceptions.CancelledError:
AttributeError: module 'asyncio' has no attribute 'exceptions'
```

File: openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py
Comment (Tilix4) [inline openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py:5]: Remove unused
```suggestion
    get_representations,
```

**Comment by kalisp** [comment]
> module 'asyncio' has no attribute 'exceptions'

That doesn't happen to me. What version of Python are you using? 3.9 as it should be on develop?

File: openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py
Comment (iLLiCiTiT) [inline openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py:123]: Representations under hero versions don't have `["context"]["version"]` so it would probably crash with `KeyError`.

File: openpype/modules/sync_server/sync_server.py
Comment (iLLiCiTiT) [inline openpype/modules/sync_server/sync_server.py:311]: Should this function copy the file to workfile directory? It looks like too much for `download_last_published_workfile`.

File: openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py
Comment (iLLiCiTiT) [reply openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py:127]: It could be like this:
```
workfile_representation  = max(
    filter(
        lambda r: r["context"].get("version") is not None,
        workfile_representations
    ),
    key=lambda d: d["context"]["version"]
)
```

Or more readable:
```
filtered_repres = filter(
    lambda r: r["context"].get("version") is not None,
    workfile_representations
)
workfile_representation  = max(filtered_repres, key=lambda r: r["context"]["version"])
```

Comment (BigRoy) [reply openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py:127]: > Could there be a hero version of workfile?

If it's a published workfile, I'd say so yes.

File: openpype/modules/sync_server/sync_server.py
Comment (iLLiCiTiT) [inline openpype/modules/sync_server/sync_server.py:317]: Also when talking about copying, this doesn't look right either. Publish and Workfile may have different versioning (there could be workfile v100 but only published workfile v1.

Comment (kalisp) [reply openpype/modules/sync_server/sync_server.py:317]: This gets triggered only if there isn't any workfile present at all. It looks only for published.

We are not syncing work workfiles. so even if there would be v100 in studio's `work`, it gets ignored.

File: openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py
Comment (kalisp) [reply openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py]: Ok, used recommended filtering.

**Comment by Tilix4** [comment]
Tested :ok:

Last thing is I'd run `black --line-length 79` to the files before merging, but this is a must have.

**Comment by kalisp** [comment]
I reformatted docs to line lenght, all others should ok.

File: website/docs/module_site_sync.md
Comment (Tilix4) [inline website/docs/module_site_sync.md:1]: I'm pretty sure this has no point to be linted. I'd `git checkout HEAD~1 website/docs/module_site_sync.md`

**Review by Tilix4 — APPROVED** [review]
LGTM, only one suggestion for markdown doc.

**Comment by iLLiCiTiT** [comment]
The prelaunch hook does not handle cases when there are no profiles in settings, or when `filter_profiles` does return `None` -> BTW I think that profiles should be empty to be skipped  by default.

**Comment by Tilix4** [comment]
Found a bug: If you delete all versions of a workfile subset for an asset and try to launch it after, it tries to find the last published workfile (which has been deleted) and hangs forever. It should not try to fetch it and open a new empty file.

**Comment by kalisp** [comment]
Fixed issue with no profiles returned.

**Comment by BigRoy** [comment]
> I tried to delete all published workfiles for particular tasks and open again, works.

How did you delete it? Did you use the loader action from the standalone loader? Because maybe @Tilix4 did use that which I believe 'archives' certain publishes instead of actually removing them from the database?

**Comment by Tilix4** [comment]
> > I tried to delete all published workfiles for particular tasks and open again, works.
> 
> How did you delete it? Did you use the loader action from the standalone loader? (OP tray > Loader) Because maybe @Tilix4 did use that which I believe 'archives' certain publishes instead of actually removing them from the database?

Indeed that's what I did. That was my guess too.

**Comment by kalisp** [comment]
Yeah, I used Delete version action in Loader.

**Comment by Tilix4** [comment]
> Yeah, I used Delete version action in Loader.

The `Delete Old Versions` one?

**Comment by kalisp** [comment]
Yeah, thats the one. I deleted all versions of "rigging" task for `Visitor` and then opened Blender without issue.
python_l3RODrvLzu

File: openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py
Comment (iLLiCiTiT) [inline openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py:160]:
```suggestion
        project_doc = self.data["project_doc"]
        project_settings = self.data["project_settings"]
```

File: openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py
Comment (iLLiCiTiT) [inline openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py:97]: This can still be used

File: openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py
Comment (iLLiCiTiT) [reply openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py:158]: Or better keep this https://github.com/ynput/OpenPype/pull/4722/files#r1168391870

File: openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py
Comment (iLLiCiTiT) [reply openpype/modules/sync_server/launch_hooks/pre_copy_last_published_workfile.py:102]: Nevermind this has app groups filtering

**Comment by Tilix4** [comment]
> Yeah, thats the one. I deleted all versions of "rigging" task for `Visitor` and then opened Blender without issue. python_l3RODrvLzu

I had this issue by publishing several versions on the same subset and deleting them. It failed until I published the same amount again and them went fine.

That being said, don't block the PR for this as it is not a very frequent use case. We could deal with it later.

**Review by iLLiCiTiT — APPROVED** [review]

**Pull request merged**
